### DISCUSSION TITLE
FDTD or BEM for scattering problems

### DISCUSSION THREAD

**Discussion by kalosu** [NanoComp/meep#2405] [Q&A]
Hello again friends from the Meep community,

I have been using meep already for a few months and I have to say that it is a really nice software package. Thanks for making it available for the general public!

I happened to come across the other project that was also developed by your group https://github.com/NanoComp/scuff-em which as far as I understand, it is used for scattering problems.

Such a situation can also be modeled using FDTD and in specific using MEEP right? Is there any situation or numerical advantage (cost of performing the computations, etc) from using SCUFF-EM?

If I want to model the case of having a complex 3D micro structure with different sub-structures(i.e, mesa structures, gratings, planar layers) being illuminated under different conditions (polarization, NA, etc) for which I am interested in computing for the field inside of my microstructure at a given point, what would be a better and "cheaper" approach for doing this? Can I use SCUFF-EM for modeling such a problem? Is MEEP a better alternative in this scenario? 

As always, thanks for your comments and feedback.

**Comment by smartalecH** [comment]
I'm sure @stevengj has a rather thorough answer for this already... but I'll throw my hat in the ring 😁

So keep in mind that the near-to-far features in meep were essentially poached from the scuff code. 

BEM codes (also called method of moments, or MoM codes) are great at modeling scattering problems for a lot of reasons. They can be super efficient, they don't suffer from the same numerical dispersion problems as finite difference methods, they (typically) incorporate a conformal mesh for the geometry, and they can be really accurate!

But BEM is a pain to set up. Every time you slightly change your problem (new material types, boundaries, etc) you often have to reformulate everything. Since it's a dense integral method (at least FEM is sparse), it can be tricky coming up with efficient solvers for different problems. The meshing itself is often a hassle. 

But FDTD is rather simple (especially in comparison)! It's pretty straightforward to augment FDTD with new features or physics. Scalability is really predictable. But of course, the standard gridding and finite differences limit the kind of problems it can solve. 

So people often marry the two techniques. Use FDTD to resolve the near fields off of some arbitrary object + source interaction, and use an integral method to compute the far fields at any arbitrary point!

Currently, meep's near2far assumes (scattered) propagation is in a uniform medium, but one could extend it to work in eg stratified media, or something even more complex (where you gradually approach a fully-featured BEM tool).  

Given your requirements, I'd probably use meep (although it sounds a bit like you're modeling a uLED-esque device... which is hard to do with *all* EM techniques...). Meep does well with complicated geometries... to a point. If you have lots of thin metals and structures that truly require a sophisticated mesh, then BEM is probably a good choice... although it's going to take some time to set up and debug. 

I'll add that scuff/buff haven't seen much love for a few years now. It's not really maintained... but it's open source so you could theoretically modify it to satisfy your needs!

**Reply by smartalecH** [reply]
(And just to add, it's hard to say if one method will be "cheaper" than the other. Usually programmer time is orders of magnitude more expensive than compute time... there are lots of considerations here)

**Comment by kalosu** [comment]
Hi @smartalecH , thanks a lot for the nice feedback and extensive explanation. 

Based on what you have described, I think the "easiest" way to start would be using meep at the moment.

For computing the scattered field I could position a DFT monitor at the points where I want to perform the evaluation by having a plane wave illumination and then run the simulation until all of the modes and transient field structure within my layers have decayed sufficiently right?

**Reply by smartalecH** [reply]
I would place dft monitors in the _near-field region_ and use a near-to-far transform (N2F) to project to the actual points in space you care about (they can be anywhere you want). As you mentioned, you need to wait for the dft fields to converge (at all frequencies you care about). You also need a closed surface (AKA a Huygen's surface).

The N2F transform is important for two reasons: (1) it's computationally more efficient because you don't have to simulate the whole domain; (2) it's more accurate because you eliminate the numerical dispersion caused by the finite differences in the FDTD algorithm.

**Reply by smartalecH** [reply]
To compute the _scattered fields_, you'll need to run two simulations: one with your structure, and one without. You'll then subtract the two field solutions (see #980 and the docs for some examples)

**Comment by stevengj** [comment]
The advantages of BEM are greatest for scattering off a small surface in a large volume, e.g. an isolated particle, especially for subwavelength metallic structures (surface plasmon resonances etcetera) where you win big by only discretizing the surface.  However, it's not easy to switch between 1d and 2d simulations, periodic boundary conditions are much more painful (but possible in SCUFF-EM), continuously varying media require a switch to VIE (implemented by BUFF-EM, but requires bigger meshes), and shape optimization is [possible](http://doi.org/10.1364/OE.397502) but requires tricky remeshing (and topology optimization is [possible](https://link.springer.com/article/10.1007/s11012-019-0095) but even harder).

Also, since it is a frequency-domain method, it has the usual pros and cons vs time-domain methods.  Pro: single-frequency or narrow-band calculations are quicker; you can input complex ε directly rather than fitting to a causal model; you can locate resonances in the complex plane to high precision using an  p (and other complex-ω tricks become easy); unstructured meshing is easier to use than in time domain.  Con: wide-band calculations can be expensive, nonlinearity or time-varying materials are hard, you have to worry more about large-scale linear-algebra techniques.
